### PR TITLE
feat(middleware): add bearer auth middleware

### LINKED ISSUE #279: Feature Request: Bearer auth
Hey! Bearer auth (eg. `Authorization: Bearer SOME_API_KEY`) is the most commonly used type of authorization header if i'm not wrong. 
It would be great to have a bearer auth middleware like [BasicAuth](https://github.com/honojs/hono/blob/master/src/middleware/basic-auth/index.ts).

I'm currently using a hardcoded version myself, so having a built-in one will help save a lot time.

Thanks

### PR DESCRIPTION
**Pull request by yusukebe**
Close #279

### CODE REVIEW

File: src/middleware/bearer-auth/index.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
+import type { Context } from '../../context'
+import type { Next } from '../../hono'
+import { timingSafeEqual } from '../../utils/buffer'
+
+const TOKEN_REGEXP = /^Bearer ([A-Za-z0-9._~+/-]+=*) *$/
```
Comment (yudai-nkt) [inline src/middleware/bearer-auth/index.ts:5]: I'm building a Discord bot with Hono, and Discord API has an Auth header in the form of <code>Authorization: Bot &#x2329;<i>token</i>&#x232A;</code> as well as the usual `Bearer` version ([reference](https://discord.com/developers/docs/reference#authentication-example-bot-token-authorization-header)). I'd be happy if this middleware supports this kind of headers out-of-the-box. Possibly via `options.prefix` parameter (defaults to `Bearer` unless explicitly specified) in `bearerAuth`? I'm not sure how many these irregular headers other than Discord API exist in the wild though.

Comment (yusukebe) [reply src/middleware/bearer-auth/index.ts]: Hi @yudai-nkt 

I didn't know `Authorization: Bot <token>`. Good suggestion! I've added the `prefix` option.
Thank you!

Comment (yudai-nkt) [reply src/middleware/bearer-auth/index.ts]: LGTM, thanks!

> I didn't know `Authorization: Bot <token>`. 

Me neither. Discord is the only example I know. 🤣

**Comment by waptik** [comment]
@yusukebe I just want to say that I tested it and it works greatly

**Pull request merged**